collector/ccr_stats.go

```diff
@@ -0,0 +1,448 @@
+					"The total time spent writing on the follower",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.TotalWriteTimeMillis)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+			{
+				Type: prometheus.CounterValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "ccr_stats", "successful_write_requests"),
+					"The number of bulk write requests executed on the follower",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.SuccessfulWriteRequests)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+			{
+				Type: prometheus.CounterValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "ccr_stats", "failed_write_requests"),
+					"The number of failed bulk write requests executed on the follower",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.FailedWriteRequests)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+			{
+				Type: prometheus.CounterValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "ccr_stats", "operations_written"),
+					"The number of operations written on the follower",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.OperationsWritten)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "ccr_stats", "time_since_last_read_millis"),
+					"The number of milliseconds since a read request was sent to the leader",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.TimeSinceListReadMillis)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "ccr_stats", "replication_lag"),
+					"Represents value of how much follower lagging the leader. Based on difference between leader_global_checkpoint and follower_global_checkpoint",
+					defaultCCRStatsLabels, nil,
+				),
+				Value: func(indexCCRStatsShard IndexCCRStatsShardResponse) float64 {
+					return float64(indexCCRStatsShard.LeaderGlobalCheckpoint - indexCCRStatsShard.FollowerGlobalCheckpoint)
+				},
+				Labels: defaultCCRLabelValues,
+			},
+		},
+	}
+}
+
+func (s *CCRStats) Describe(ch chan<- *prometheus.Desc) {
+	for _, metric := range s.ccrMetrics {
+		ch <- metric.Desc
+	}
+	ch <- s.up.Desc()
+	ch <- s.totalScrapes.Desc()
+	ch <- s.jsonParseFailures.Desc()
+}
+
+func (s *CCRStats) fetchCCRStatsResponse() (CCRStatsResponse, error) {
```

Comment:
All of this logic can go into the Update func. No need for a dedicated function.